Update build scripts

stvstnfrd opened the pull request:
None of these appear to be used; some are redundant.

The rename/relocation of the "create XBlock" script serves a few purposes..

Following *nix convention, you can add the `bin` directory to your `${PATH}` to make the script directly invokable. Given this, the `workbench-` prefix is important  as it removes ambiguity from the command name; `startnew.py` is a poor name when exposed "globally".

If all new scripts (yes, expect additional PRs shortly) follow this convention, all workbench commands are "grouped"; auto/tab-completion FTW.

edx-webhook commented:
Thanks for the pull request, @stvstnfrd! I've created [OSPR-855](https://openedx.atlassian.net/browse/OSPR-855) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

sarina commented:
Instead of removing the quality script can we make Travis run it?

sarina commented:
and I guess we don't need `coverage` w/ coveralls

stvstnfrd commented:
@sarina I added a new target `make quality` which runs PEP8 and Pylint. This is included as part of Travis' build. Consequently, a number of quality fixes are now included as separate commits.

stvstnfrd commented:
Disclaimer: Coveralls is reporting a minor (0.02%) drop in coverage.

Upon inspection, this appears to be an "accounting mistake"; some of the PEP8/Pylint refactoring resulted in a slightly higher line count amongst untested code paths. So the logic and its testing are unchanged, but _technically_ there are more untested lines. I'd like to consider this a non-issue and ignore during merge consideration.

sarina commented:
Yeah that makes sense to me. I can take a review pass this week; @singingwolfboy can you review too?

stvstnfrd commented on pylintrc:
```diff
@@ -34,6 +34,9 @@ load-plugins=
 # multiple time (only on the command line, not in the configuration file where
 # it should appear only once).
 disable=
+    missing-docstring,
```
This is by far the worst offender. Eventually it will be nice to round out the documentation, but for now, I'd consider this outside the scope. We can add additional docstrings in a future pass, then remove this line.

sarina replied:
I feel it's more visible to keep this in, but set a threshold to account for how many docstrings are currently missing. Also, it means that new code isn't held to the higher standard, which I'm not a fan of.

singingwolfboy replied:
I agree with @sarina. It's useful to have pylint flag new functions that don't have docstrings.

stvstnfrd replied:
I agree about the downside of holding new code to the higher standard, but disagree about the utility of displaying "old" violations.

In this case, there are more than 2 dozen such violations. So as a developer, if I get a failing quality build, I now have to wade through twenty-some violations and try to figure which ones are worth paying attention to. I know this has previously been a pain point in the platform.

Subsequently, I just went ahead and added the missing docstrings, so it's a moot point now.

sarina replied:
> I now have to wade through twenty-some violations

Well, this is in part one of the advantages of diff-quality, but as you
say, moot point now

On Mon, Oct 19, 2015 at 5:03 PM, Steven Burch <notifications@github.com>
wrote:

> In pylintrc
> <https://github.com/edx/xblock-sdk/pull/71#discussion_r42425523>:
>
> > @@ -34,6 +34,9 @@ load-plugins=
> >  # multiple time (only on the command line, not in the configuration file where
> >  # it should appear only once).
> >  disable=
> > +    missing-docstring,
>
> I agree about the downside of holding new code to the higher standard, but
> disagree about the utility of displaying "old" violations.
>
> In this case, there are more than 2 dozen such violations. So as a
> developer, if I get a failing quality build, I now have to wade through
> twenty-some violations and try to figure which ones are worth paying
> attention to. I know this has previously been a pain point in the platform.
>
> Subsequently, I just went ahead and added the missing docstrings, so it's
> a moot point now.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/edx/xblock-sdk/pull/71/files#r42425523>.
>

stvstnfrd commented on pylintrc:
```diff
@@ -34,6 +34,9 @@ load-plugins=
 # it should appear only once).
 disable=
+    missing-docstring,
+    fixme,
```
Useful to know, but not worth failing builds over.

singingwolfboy commented on sample_xblocks/filethumbs/filethumbs.py:
```diff
@@ -18,14 +19,18 @@
-         "11011",
-         "11011",
-         "10001"]
-arrow = [map(int, x) for x in arrow]
```
Fascinating. This obscure piece of code seems to be just converting strings to ints. Why not define the list with ints to begin with?

stvstnfrd replied:
Maybe, but I'm not looking to refactor the entire codebase at this point, so I'm considering it outside the scope of the submission. I'm only touching this to eliminate quality violations.

singingwolfboy replied:
Makes sense.

singingwolfboy commented on sample_xblocks/filethumbs/filethumbs.py:
```diff
@@ -59,9 +64,9 @@ def student_view(self, context=None):  # pylint: disable=W0613
-                f.close()
+            with self.fs.open('thumbsvotes.json', 'wb') as file_output:
+                json.dump({'up': 0, 'down': 0}, file_output)
+                file_output.close()
```
I'm fine with changing `f` to `file_output`, but personally, I'm also OK with just marking `f` as an acceptable variable name in the `pylintrc` file. I think that `f` is a clear name in context, especially since it's widely used for this purpose, and the shortness of the name indicates that the variable is short-lived.

stvstnfrd replied:
I disagree about pylintrc. By adding the exception, we enable developers to use more ambiguous variable names elsewhere in the code.

Also, `f` is objectively more ambiguous than `file_output`. The latter provides additional context about the file pointer and its behavior; I know what I can do with the object without inspecting its initialization.

Despite frequent arguments from developers that truncated variable names "save time typing" or some such nonsense, it's important to remember that all code will be read _many_ more times than it is written. We should _always_ favor explicit naming conventions; it's a no-lose situation.

sarina replied:
Throwing in my chips with @stvstnfrd here - better var names keep on the path to self documenting code.

singingwolfboy commented on sample_xblocks/filethumbs/filethumbs.py:
```diff
@@ -115,9 +120,8 @@ def vote(self, data, suffix=''):  # pylint: disable=unused-argument
-        json.dump({'up': self.upvotes, 'down': self.downvotes}, f)
-        f.close()
+        with self.fs.open('thumbsvotes.json', 'wb') as file_output:
+            json.dump({'up': self.upvotes, 'down': self.downvotes}, file_output)
```
Yay for context-manager form! Much clearer, and less prone to errors. :dancer:

singingwolfboy commented on setup.py:
```diff
@@ -65,7 +65,14 @@ def load_requirements(*requirements_paths):
     ],
     install_requires=load_requirements('requirements.txt'),
-    tests_require=load_requirements('test-requirements.txt'),
+    tests_require=[
```
I'm fine with deleting `test-requirements.txt` and putting that info directly into the `setup.py` file, but I'm curious why you want to do so -- especially since you're leaving `requirements.txt` unchanged. What's the difference?

stvstnfrd replied:
This was legacy cargo cult. External requirement files were needed to install packages that weren't on PyPi (i.e. Github). All of the test packages are on PyPi, so we can remove the indirection. We don't gain anything from the extra file; `setup.py` is already just reading in its contents anyway.

`requirements.txt` is still installing XBlock via Github, so I'm considering that change outside the scope of this submission.

stvstnfrd replied:
Ugh.. this was one of those "last-minute, what could it hurt" sort of changes, which, well... never end up being so straight-foward.  There are some additional caveats to this change, namely, `tests_require` doesn't actually install the packages into the `virtualenv`; it just installs them while running the tests.

While I'd like to revisit this in the future, I want to avoid rabbit-holing until I can devote time to straightening out all of the plumbing.

For the sake of this PR, I have reverted the change.

singingwolfboy commented on workbench/runtime.py:
```diff
@@ -201,7 +201,7 @@ def get_usage_id_from_aside(self, aside_id):
             return self._aside_usages[aside_id][0]
         except KeyError:
-            raise NoSuchUsage(usage_id)
+            raise NoSuchUsage(aside_id)
```
This seems like a larger change. I don't know if it's correct, but I wanted to flag it and ask someone else who might know the difference to review it.

stvstnfrd replied:
> This seems like a larger change.

It's not; this is me fixing a copy/paste fail.

If that line is ever run, it would break because `usage_id` is never defined in this context.

This method and several surrounding it (`get_usage_id_from_aside` and `get_aside_id_from_usage`, etc.) are nearly identical. Upon inspection, they reveal the form:

```python
def get_THING_id_from_STUFF(self, STUFF_ID):
    try:
        return self.aside_usages[STUFF_ID][INDEX]
    except KeyError:
        raise NoSuchUsage(STUFF_ID)
```

I'd bet dollars to donuts that `get_aside_type_from_usage` was written first, then copy/pasted and then they forgot to edit the line in the exception handler. Odds are either this exception is never encountered, or people just don't notice because it ends up just displaying a different kind of exception.

CC: @cpennington (original author)

cpennington replied:
:+1:

singingwolfboy commented on workbench/test/test_problems.py:
```diff
@@ -60,7 +60,12 @@ def assert_image(right_wrong_idx, expected_icon):
+                        css='{} img'.format(
+                            check_indicators,
+                        ),
+                    ).nth(right_wrong_idx).attrs('src')
```
Much clearer! Thank you! :smile:

singingwolfboy commented on workbench/views.py:
```diff
@@ -80,14 +79,20 @@ def show_scenario(request, scenario_id, view_name='student_view', template='work
 
 
-def user_list(request):
+def user_list(_request):
```
If we're not using the `request` parameter, why even include it at all? Why not remove it?

singingwolfboy replied:
To clarify, this change actually changes the signature of this function if the caller expects to pass this argument as a keyword: `user_list(request=foo)`. We should only make this change if we're sure that no caller does this. If we think that a caller might do this, then it would be better to just use a `pylint: disable` here.

stvstnfrd replied:
`user_list` is actually a view defined in `workbench/urls.py`, so the `_request` object will always be passed along.

stvstnfrd replied:
To explicitly address your follow-up point, I don't think we need to be concerned; this is a Django View, not intended to be called directly by developers.

stvstnfrd commented:
@sarina @singingwolfboy Thanks for the quick review! I have added clarifying comments where requested and made the following significant changes:
- removed ignoring of `missing-docstring` and just documented them
- removed `test_requires` changes; deferring until a later date.

Let me know if you need anything else; otherwise this is ready to merge on my end.

sarina commented:
:+1: 

I'll let @singingwolfboy do final review & merge, but this is good by me. DB - the coverage slight drop is expected, and I assert we should accept.
